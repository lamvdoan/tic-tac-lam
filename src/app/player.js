export default class Player {
    constructor(marker, playerId) {
        this.marker = marker;
        this.playerId = playerId;
    }

    getMarker() {
        return this.marker;
    }

    getPlayerId() {
        return this.playerId;
    }
}
