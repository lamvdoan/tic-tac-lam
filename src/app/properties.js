const UNOCCUPIED_CELL = '';
const MAX_NUMBER_OF_CELLS = 9;
const PLAYER_ONE_ID = 1;
const PLAYER_TWO_ID = 2;
const WINNING_CELLS_LIST = [
    [0, 1, 2],
    [3, 4, 5],
    [6, 7, 8],
    [0, 3, 6],
    [1, 4, 7],
    [2, 5, 8],
    [0, 4, 8],
    [2, 4, 6]
];

export { UNOCCUPIED_CELL, MAX_NUMBER_OF_CELLS, PLAYER_ONE_ID, PLAYER_TWO_ID, WINNING_CELLS_LIST } ;