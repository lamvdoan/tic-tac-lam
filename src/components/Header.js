import React from 'react';
import { Link } from 'react-router-dom';

const Header = () => (
    <header className="header">
        <div className="content-container">
            <Link className="header__title" to="/">
                <h1>Tic Tac Lam</h1>
            </Link>
        </div>
    </header>
);
export default Header;
