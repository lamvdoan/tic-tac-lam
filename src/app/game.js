import { Player } from './player';
import { UNOCCUPIED_CELL, MAX_NUMBER_OF_CELLS, PLAYER_ONE_ID, PLAYER_TWO_ID, WINNING_CELLS_LIST } from './properties';

export class Game {
    constructor() {
        this.players = [new Player('X', PLAYER_ONE_ID), new Player('O', PLAYER_TWO_ID)];
    }

    startGame() {
        this.grid = Array(MAX_NUMBER_OF_CELLS).fill(UNOCCUPIED_CELL);
        this.isGameActive = true;
        this.currentPlayer = this.players[0];
        console.log('Game starting...');
    }

    chooseACell(index) {
        if (this.isGameActive) {
            if (this.isMoveValid(index)) {
                this.cellIsMarked(index);
                this.checkForWinCondition();
                this.checkForCatCondition();
                this.nextPlayer();
            } else {
                // Display message to try again
            }
        }
    }

    isMoveValid(index) {
        return (this.grid[index] === UNOCCUPIED_CELL) ? true : false;
    }

    cellIsMarked(index) {
        this.grid[index] = this.currentPlayer.getMarker();
    }

    checkForWinCondition() {
        const getPlayerMarkedCellsReducer = (accumulator, cell, index) => {
            if (cell === this.currentPlayer.getMarker()) {
                accumulator.push(index);
            }

            return accumulator;
        };

        const playerCells = this.grid.reduce(getPlayerMarkedCellsReducer, []);

        for (let i = 0; i < WINNING_CELLS_LIST.length; i++) {
            let winningCells = Array.from(WINNING_CELLS_LIST[i]);
            let didPlayerWin = winningCells.every(cell => playerCells.includes(cell));

            if (didPlayerWin) {
                this.isGameActive = false;
                break;

            // display victory for specified player
            }
        }
    }

    checkForCatCondition() {
        if (this.isGameActive) {
            const countOfOccupiedCells = this.grid.filter(cell => cell != UNOCCUPIED_CELL);

            if (countOfOccupiedCells.length == MAX_NUMBER_OF_CELLS) {
                this.isGameActive = false;

                // display cat condition
            }
        }
    }

    nextPlayer() {
        if (this.isGameActive) {
            this.currentPlayer = (this.currentPlayer.getPlayerId() === PLAYER_ONE_ID) ? this.players[1] : this.players[0];

            // display whose turn
        }
    }
}
