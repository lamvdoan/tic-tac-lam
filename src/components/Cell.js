import React from 'react';
import { connect } from 'react-redux';
import { cellIsMarked, setIsGameActive, setCurrentPlayer } from '../actions/game';
import { O_CELL_IMAGE, X_CELL_IMAGE, EMPTY_CELL_IMAGE, IMAGE_HOME } from '../properties/imageFileSystem';

const imagePath = {
    emptyCell: `${IMAGE_HOME}/${EMPTY_CELL_IMAGE}`,
    oCell: `${IMAGE_HOME}/${O_CELL_IMAGE}`,
    xCell: `${IMAGE_HOME}/${X_CELL_IMAGE}`,
}

export class Cell extends React.Component {
    onClick = () => {
        console.log('Clicking cell: ' + this.props.id);
        this.props.cellIsMarked(this.props.id);
        this.props.setCurrentPlayer();

        // console.log('Grid Letter: ' + this.props.game.grid[this.props.id]);
        // console.log('Grid: ' + this.props.game.grid);
    };

    getImageName = () => {
        if (this.props.game.grid[this.props.id] === 'X') {
            return 'xCell';
        } else if (this.props.game.grid[this.props.id] === 'O') {
            return 'oCell';
        }
    }

    render() {
        const imageName = this.getImageName();
        return (
            <img onClick={this.onClick} className="cell" src={imagePath[imageName]} />
        );
    }
}

const mapStateToProps = (state) => ({
    game: state.game
});

const mapDispatchToProps = (dispatch) => ({
    cellIsMarked: (index) => dispatch(cellIsMarked(index)),
    setIsGameActive: () => dispatch(setIsGameActive()),
    setCurrentPlayer: () => dispatch(setCurrentPlayer())
});

export default connect(mapStateToProps, mapDispatchToProps)(Cell);
