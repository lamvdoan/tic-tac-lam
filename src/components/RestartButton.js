import React from 'react';
import { connect } from 'react-redux';
import { restartGame } from '../actions/game';

export class RestartButton extends React.Component {
    onClick = () => {
        this.props.restartGame();
    }

    render() {
        return (
            <button onClick={this.onClick} className="button">Play Again</button>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    restartGame: () => dispatch(restartGame())
});

export default connect(null, mapDispatchToProps)(RestartButton);
