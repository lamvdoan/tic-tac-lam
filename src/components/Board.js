import React from 'react';
import Cell from './Cell';

export class Board extends React.Component {
    render() {
        return (
            <div className="board">
                <div className="board__row">
                    <Cell id="0" />
                    <Cell id="1"/>
                    <Cell id="2"/>
                </div>
                <div className="board__row">
                    <Cell id="3"/>
                    <Cell id="4"/>
                    <Cell id="5"/>
                </div>
                <div className="board__row">
                    <Cell id="6"/>
                    <Cell id="7"/>
                    <Cell id="8"/>
                </div>
            </div>
        );
    }
}

export default Board;
