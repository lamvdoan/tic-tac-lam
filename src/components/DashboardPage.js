import React from 'react';
import Board from './Board';

const DashboardPage = () => (
    <div>
        <Board />
    </div>
);

export default DashboardPage;