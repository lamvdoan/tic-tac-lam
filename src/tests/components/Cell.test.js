import React from 'react';
import { shallow } from 'enzyme';
import { Cell } from '../../components/Cell';
import { X_CELL, O_CELL, UNOCCUPIED_CELL } from '../../app/properties';

let game;
let chooseACell;

beforeEach(() => {
    game = {
        grid: [X_CELL, O_CELL, UNOCCUPIED_CELL]
    };
    chooseACell = jest.fn();
});

test('should render cell', () => {
    let id = 0;
    let wrapper = shallow(<Cell chooseACell={chooseACell} id={id} game={game}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should handle onClick', () => {
    let id = 0;
    let wrapper = shallow(<Cell chooseACell={chooseACell} id={id} game={game}/>);
    wrapper.find('div').simulate('click');
    expect(chooseACell).toHaveBeenLastCalledWith(id);
});

test('should get X marker', () => {
    let id = 0;
    let wrapper = shallow(<Cell chooseACell={chooseACell} id={id} game={game}/>);
    expect(wrapper.instance().getMarker()).toEqual(X_CELL);
});

test('should get O marker', () => {
    let id = 1;
    let wrapper = shallow(<Cell chooseACell={chooseACell} id={id} game={game}/>);
    expect(wrapper.instance().getMarker()).toEqual(O_CELL);
});

test('should get unoccupied marker', () => {
    let id = 2;
    let wrapper = shallow(<Cell chooseACell={chooseACell} id={id} game={game}/>);
    expect(wrapper.instance().getMarker()).toEqual(UNOCCUPIED_CELL);
});
