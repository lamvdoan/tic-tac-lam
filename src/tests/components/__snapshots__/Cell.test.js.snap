// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render cell 1`] = `
<div
  className="cell"
  onClick={[Function]}
>
  X
</div>
`;
